commit 6f75707f34c3b081e8c7b9ba96031bf016f98816
Author:     Yuhong Wen <yuhongw@nvidia.com>
AuthorDate: Tue Feb 21 14:01:34 2023 -0500
Commit:     GitHub <noreply@github.com>
CommitDate: Tue Feb 21 14:01:34 2023 -0500

    Job run integrate with FCI Cellnet (#1387)
    
    * FL server integrate with FCI cellnet.
    
    * Codestyle reformat.
    
    * fed_server_test.py integrate with Cellnet.
    
    * FL client integrate with Cellnet. Client register to server.
    
    * Reformat codestyle.
    
    * reformat codestyle.
    
    * update the message.
    
    * Codestyle reformat.
    
    * Fixed the import sort.
    
    * Addes the PR reviews.
    
    * codestyle fix.
    
    * made the cell_timeout configurable.
    
    * codestyle fix.
    
    * removed no use import.
    
    * disable simulator_runner_test temporary.
    
    * FCI integration for job run.
    
    * Fix for the admin auto login.
    
    * reformat codestyle.
    
    * moved create_admin_server() to utils.
    
    * removed the no use import.
    
    * sort import.
    
    * Changes after the PR reviews.
    
    * rolled back the change dh_psi_test.py.
    
    * Removed no use import.
    
    * PR review changes.
    
    * type hint change.
---
 nvflare/app_common/job_schedulers/job_scheduler.py |   2 +-
 nvflare/private/fed/client/client_engine.py        |   1 -
 nvflare/private/fed/server/job_runner.py           | 176 +++++------
 nvflare/private/fed/server/server_engine.py        | 324 ++++++---------------
 4 files changed, 188 insertions(+), 315 deletions(-)

diff --git a/nvflare/app_common/job_schedulers/job_scheduler.py b/nvflare/app_common/job_schedulers/job_scheduler.py
index 17f0617e..cfb89eac 100644
--- a/nvflare/app_common/job_schedulers/job_scheduler.py
+++ b/nvflare/app_common/job_schedulers/job_scheduler.py
@@ -155,7 +155,7 @@ class DefaultJobScheduler(JobSchedulerSpec, FLComponent):
         job_participants = [fl_ctx.get_identity_name(default="server")]
         job_participants.extend(applicable_sites)
 
-        fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, job.job_id, private=True, sticky=False)
+        fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, job.job_id, private=True)
         fl_ctx.set_prop(FLContextKey.CLIENT_RESOURCE_SPECS, resource_reqs, private=True, sticky=False)
         fl_ctx.set_prop(FLContextKey.JOB_PARTICIPANTS, job_participants, private=True, sticky=False)
         self.fire_event(EventType.BEFORE_CHECK_CLIENT_RESOURCES, fl_ctx)
diff --git a/nvflare/private/fed/client/client_engine.py b/nvflare/private/fed/client/client_engine.py
index dced5698..da781d20 100644
--- a/nvflare/private/fed/client/client_engine.py
+++ b/nvflare/private/fed/client/client_engine.py
@@ -53,7 +53,6 @@ class ClientEngine(ClientEngineInternalSpec):
         Args:
             client: FL client object
             client_name: client name
-            sender: sender object
             args: command args
             rank: local process rank
             workers: number of workers
diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index aa259092..b7ba2127 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -36,7 +36,7 @@ from nvflare.security.logging import secure_format_exception
 
 
 def _send_to_clients(admin_server, client_sites: List[str], engine, message):
-    clients, invalid_inputs = engine.validate_clients(client_sites)
+    clients, invalid_inputs = engine.validate_targets(client_sites)
     if invalid_inputs:
         raise RuntimeError(f"invalid clients: {invalid_inputs}.")
     requests = {}
@@ -143,7 +143,7 @@ class JobRunner(FLComponent):
 
             if client_sites:
                 message = self._make_deploy_message(job, app_data, app_name)
-                clients, invalid_inputs = engine.validate_clients(client_sites)
+                clients, invalid_inputs = engine.validate_targets(client_sites)
 
                 if invalid_inputs:
                     deploy_detail.append("invalid_clients: {}".format(",".join(invalid_inputs)))
@@ -293,25 +293,26 @@ class JobRunner(FLComponent):
         while not self.ask_to_stop:
             for job_id in list(self.running_jobs.keys()):
                 if job_id not in engine.run_processes.keys():
-                    with self.lock:
-                        job = self.running_jobs.get(job_id)
-                        if job:
-                            exception_run_processes = engine.exception_run_processes
-                            if job_id in exception_run_processes:
-                                self.log_info(fl_ctx, f"Try to abort run ({job_id}) on clients.")
-                                run_process = exception_run_processes[job_id]
-
-                                # stop client run
-                                client_sites = run_process.get(RunProcessKey.PARTICIPANTS)
-                                self.abort_client_run(job_id, client_sites, fl_ctx)
-
-                                job_manager.set_status(job.job_id, RunStatus.FINISHED_EXECUTION_EXCEPTION, fl_ctx)
-                            else:
-                                job_manager.set_status(job.job_id, RunStatus.FINISHED_COMPLETED, fl_ctx)
+                    # with self.lock:
+                    job = self.running_jobs.get(job_id)
+                    if job:
+                        exception_run_processes = engine.exception_run_processes
+                        if job_id in exception_run_processes:
+                            self.log_info(fl_ctx, f"Try to abort run ({job_id}) on clients.")
+                            run_process = exception_run_processes[job_id]
+
+                            # stop client run
+                            client_sites = run_process.get(RunProcessKey.PARTICIPANTS)
+                            self.abort_client_run(job_id, client_sites, fl_ctx)
+
+                            job_manager.set_status(job.job_id, RunStatus.FINISHED_EXECUTION_EXCEPTION, fl_ctx)
+                        else:
+                            job_manager.set_status(job.job_id, RunStatus.FINISHED_COMPLETED, fl_ctx)
+                        with self.lock:
                             del self.running_jobs[job_id]
-                            fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, job.job_id)
-                            self.fire_event(EventType.JOB_COMPLETED, fl_ctx)
-                            self.log_debug(fl_ctx, f"Finished running job:{job.job_id}")
+                        fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, job.job_id)
+                        self.fire_event(EventType.JOB_COMPLETED, fl_ctx)
+                        self.log_debug(fl_ctx, f"Finished running job:{job.job_id}")
                     engine.remove_exception_process(job_id)
             time.sleep(1.0)
 
@@ -344,67 +345,67 @@ class JobRunner(FLComponent):
                     )
 
                     if ready_job:
-                        with self.lock:
-                            client_sites = {k: v for k, v in sites.items() if k != "server"}
-                            job_id = None
-                            try:
-                                self.log_info(fl_ctx, f"Got the job:{ready_job.job_id} from the scheduler to run")
-                                fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, ready_job.job_id)
-                                job_id, failed_clients = self._deploy_job(ready_job, sites, fl_ctx)
-                                job_manager.set_status(ready_job.job_id, RunStatus.DISPATCHED, fl_ctx)
-
-                                deploy_detail = fl_ctx.get_prop(FLContextKey.JOB_DEPLOY_DETAIL)
-                                if deploy_detail:
-                                    job_manager.update_meta(
-                                        ready_job.job_id,
-                                        {
-                                            JobMetaKey.JOB_DEPLOY_DETAIL.value: deploy_detail,
-                                            JobMetaKey.SCHEDULE_COUNT.value: ready_job.meta[
-                                                JobMetaKey.SCHEDULE_COUNT.value
-                                            ],
-                                            JobMetaKey.LAST_SCHEDULE_TIME.value: ready_job.meta[
-                                                JobMetaKey.LAST_SCHEDULE_TIME.value
-                                            ],
-                                            JobMetaKey.SCHEDULE_HISTORY.value: ready_job.meta[
-                                                JobMetaKey.SCHEDULE_HISTORY.value
-                                            ],
-                                        },
-                                        fl_ctx,
-                                    )
-
-                                if failed_clients:
-                                    deployable_clients = {
-                                        k: v for k, v in client_sites.items() if k not in failed_clients
-                                    }
-                                else:
-                                    deployable_clients = client_sites
-
-                                self._start_run(
-                                    job_id=job_id,
-                                    job=ready_job,
-                                    client_sites=deployable_clients,
-                                    fl_ctx=fl_ctx,
+                        # with self.lock:
+                        client_sites = {k: v for k, v in sites.items() if k != "server"}
+                        job_id = None
+                        try:
+                            self.log_info(fl_ctx, f"Got the job:{ready_job.job_id} from the scheduler to run")
+                            fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, ready_job.job_id)
+                            job_id, failed_clients = self._deploy_job(ready_job, sites, fl_ctx)
+                            job_manager.set_status(ready_job.job_id, RunStatus.DISPATCHED, fl_ctx)
+
+                            deploy_detail = fl_ctx.get_prop(FLContextKey.JOB_DEPLOY_DETAIL)
+                            if deploy_detail:
+                                job_manager.update_meta(
+                                    ready_job.job_id,
+                                    {
+                                        JobMetaKey.JOB_DEPLOY_DETAIL.value: deploy_detail,
+                                        JobMetaKey.SCHEDULE_COUNT.value: ready_job.meta[
+                                            JobMetaKey.SCHEDULE_COUNT.value
+                                        ],
+                                        JobMetaKey.LAST_SCHEDULE_TIME.value: ready_job.meta[
+                                            JobMetaKey.LAST_SCHEDULE_TIME.value
+                                        ],
+                                        JobMetaKey.SCHEDULE_HISTORY.value: ready_job.meta[
+                                            JobMetaKey.SCHEDULE_HISTORY.value
+                                        ],
+                                    },
+                                    fl_ctx,
                                 )
+
+                            if failed_clients:
+                                deployable_clients = {k: v for k, v in client_sites.items() if k not in failed_clients}
+                            else:
+                                deployable_clients = client_sites
+
+                            self._start_run(
+                                job_id=job_id,
+                                job=ready_job,
+                                client_sites=deployable_clients,
+                                fl_ctx=fl_ctx,
+                            )
+                            with self.lock:
                                 self.running_jobs[job_id] = ready_job
-                                job_manager.set_status(ready_job.job_id, RunStatus.RUNNING, fl_ctx)
-                            except BaseException as e:
-                                if job_id:
-                                    if job_id in self.running_jobs:
+                            job_manager.set_status(ready_job.job_id, RunStatus.RUNNING, fl_ctx)
+                        except BaseException as e:
+                            if job_id:
+                                if job_id in self.running_jobs:
+                                    with self.lock:
                                         del self.running_jobs[job_id]
-                                    self._stop_run(job_id, fl_ctx)
-                                job_manager.set_status(ready_job.job_id, RunStatus.FAILED_TO_RUN, fl_ctx)
-
-                                deploy_detail = fl_ctx.get_prop(FLContextKey.JOB_DEPLOY_DETAIL)
-                                if deploy_detail:
-                                    job_manager.update_meta(
-                                        ready_job.job_id, {JobMetaKey.JOB_DEPLOY_DETAIL.value: deploy_detail}, fl_ctx
-                                    )
-
-                                self.fire_event(EventType.JOB_ABORTED, fl_ctx)
-                                self.log_error(
-                                    fl_ctx, f"Failed to run the Job ({ready_job.job_id}): {secure_format_exception(e)}"
+                                self._stop_run(job_id, fl_ctx)
+                            job_manager.set_status(ready_job.job_id, RunStatus.FAILED_TO_RUN, fl_ctx)
+
+                            deploy_detail = fl_ctx.get_prop(FLContextKey.JOB_DEPLOY_DETAIL)
+                            if deploy_detail:
+                                job_manager.update_meta(
+                                    ready_job.job_id, {JobMetaKey.JOB_DEPLOY_DETAIL.value: deploy_detail}, fl_ctx
                                 )
 
+                            self.fire_event(EventType.JOB_ABORTED, fl_ctx)
+                            self.log_error(
+                                fl_ctx, f"Failed to run the Job ({ready_job.job_id}): {secure_format_exception(e)}"
+                            )
+
                 time.sleep(1.0)
         else:
             self.log_error(fl_ctx, "There's no Job Manager defined. Won't be able to run the jobs.")
@@ -430,17 +431,20 @@ class JobRunner(FLComponent):
     def stop_run(self, job_id: str, fl_ctx: FLContext):
         engine = fl_ctx.get_engine()
         job_manager = engine.get_component(SystemComponents.JOB_MANAGER)
-        with self.lock:
-            self._stop_run(job_id, fl_ctx)
-            job = self.running_jobs.get(job_id)
-            if job:
-                self.log_info(fl_ctx, f"Stop the job run: {job_id}")
-                fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, job.job_id)
-                job_manager.set_status(job.job_id, RunStatus.FINISHED_ABORTED, fl_ctx)
+        self._stop_run(job_id, fl_ctx)
+
+        # with self.lock:
+        job = self.running_jobs.get(job_id)
+        if job:
+            self.log_info(fl_ctx, f"Stop the job run: {job_id}")
+            fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, job.job_id)
+            job_manager.set_status(job.job_id, RunStatus.FINISHED_ABORTED, fl_ctx)
+            with self.lock:
                 del self.running_jobs[job_id]
-                self.fire_event(EventType.JOB_ABORTED, fl_ctx)
-            else:
-                raise RuntimeError(f"Job {job_id} is not running.")
+            self.fire_event(EventType.JOB_ABORTED, fl_ctx)
+        else:
+            # raise RuntimeError(f"Job {job_id} is not running.")
+            self.log_error(fl_ctx, f"Job {job_id} is not running.")
 
     def stop_all_runs(self, fl_ctx: FLContext):
         engine = fl_ctx.get_engine()
diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index 5e2e407a..b7dc502d 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -23,8 +23,6 @@ import sys
 import threading
 import time
 from concurrent.futures import ThreadPoolExecutor
-from multiprocessing.connection import Client as CommandClient
-from multiprocessing.connection import Listener
 from threading import Lock
 from typing import Dict, List, Optional, Tuple
 
@@ -34,7 +32,6 @@ from nvflare.apis.fl_constant import (
     AdminCommandNames,
     FLContextKey,
     MachineStatus,
-    ReservedTopic,
     ReturnCode,
     RunProcessKey,
     ServerCommandKey,
@@ -46,14 +43,16 @@ from nvflare.apis.fl_context import FLContext, FLContextManager
 from nvflare.apis.fl_snapshot import RunSnapshot
 from nvflare.apis.impl.job_def_manager import JobDefManagerSpec
 from nvflare.apis.job_def import Job
-from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.shareable import Shareable
 from nvflare.apis.utils.fl_context_utils import get_serializable_data
 from nvflare.apis.workspace import Workspace
+from nvflare.fuel.f3.cellnet.cell import FQCN, Cell
 from nvflare.fuel.utils import fobs
+from nvflare.fuel.utils.argument_utils import parse_vars
 from nvflare.fuel.utils.network_utils import get_open_ports
 from nvflare.fuel.utils.zip_utils import zip_directory_to_bytes
 from nvflare.private.admin_defs import Message, MsgHeader
-from nvflare.private.defs import RequestHeader, TrainingTopic
+from nvflare.private.defs import CellChannel, CellMessageHeaderKeys, RequestHeader, TrainingTopic, new_cell_message
 from nvflare.private.fed.server.server_json_config import ServerJsonConfigurator
 from nvflare.private.fed.utils.fed_utils import security_close
 from nvflare.private.scheduler_constants import ShareableHeader
@@ -61,11 +60,11 @@ from nvflare.security.logging import secure_format_exception
 from nvflare.widgets.info_collector import InfoCollector
 from nvflare.widgets.widget import Widget, WidgetID
 
-from .admin import ClientReply
 from .client_manager import ClientManager
 from .job_runner import JobRunner
-from .run_manager import RunInfo, RunManager
-from .server_commands import NO_OP_REPLY
+from .message_send import ClientReply
+from .run_info import RunInfo
+from .run_manager import RunManager
 from .server_engine_internal_spec import EngineInfo, ServerEngineInternalSpec
 from .server_status import ServerStatus
 
@@ -118,11 +117,11 @@ class ServerEngine(ServerEngineInternalSpec):
 
         self.asked_to_stop = False
         self.snapshot_persistor = snapshot_persistor
-        self.parent_conn = None
-        self.parent_conn_lock = Lock()
         self.job_runner = None
         self.job_def_manager = None
 
+        self.kv_list = parse_vars(args.set)
+
     def _get_server_app_folder(self):
         return WorkspaceConstants.APP_PREFIX + "server"
 
@@ -157,35 +156,6 @@ class ServerEngine(ServerEngineInternalSpec):
             return run_info
         return None
 
-    def create_parent_connection(self, port):
-        while not self.parent_conn:
-            try:
-                address = ("localhost", port)
-                self.parent_conn = CommandClient(address, authkey="parent process secret password".encode())
-            except BaseException:
-                time.sleep(1.0)
-                pass
-
-        threading.Thread(target=self.heartbeat_to_parent, args=[]).start()
-
-    def heartbeat_to_parent(self):
-        while True:
-            try:
-                with self.parent_conn_lock:
-                    data = {ServerCommandKey.COMMAND: ServerCommandNames.HEARTBEAT, ServerCommandKey.DATA: {}}
-                    self.parent_conn.send(data)
-                time.sleep(1.0)
-            except BaseException:
-                # The parent process can not be reached. Terminate the child process.
-                break
-        # wait and delay some time for the wrap up process before the child process self terminate.
-        start_time = time.time()
-        while self.engine_info.status != MachineStatus.STOPPED:
-            time.sleep(1.0)
-            if time.time() - start_time > 30.0:
-                break
-        os.killpg(os.getpgid(os.getpid()), 9)
-
     def delete_job_id(self, num):
         job_id_folder = os.path.join(self.args.workspace, WorkspaceConstants.WORKSPACE_PREFIX + str(num))
         if os.path.exists(job_id_folder):
@@ -195,8 +165,8 @@ class ServerEngine(ServerEngineInternalSpec):
     def get_clients(self) -> [Client]:
         return list(self.client_manager.get_clients().values())
 
-    def validate_clients(self, client_names: List[str]) -> Tuple[List[Client], List[str]]:
-        return self._get_all_clients_from_inputs(client_names)
+    def validate_targets(self, client_names: List[str]) -> Tuple[List[Client], List[str]]:
+        return self.client_manager.get_all_clients_from_inputs(client_names)
 
     def start_app_on_server(self, run_number: str, job: Job = None, job_clients=None, snapshot=None) -> str:
         if run_number in self.run_processes.keys():
@@ -215,76 +185,38 @@ class ServerEngine(ServerEngineInternalSpec):
 
             open_ports = get_open_ports(2)
             self._start_runner_process(
-                self.args, app_root, run_number, app_custom_folder, open_ports, job.job_id, job_clients, snapshot
+                self.args,
+                app_root,
+                run_number,
+                app_custom_folder,
+                open_ports,
+                job.job_id,
+                job_clients,
+                snapshot,
+                self.server.cell,
             )
 
-            threading.Thread(target=self._listen_command, args=(open_ports[0], run_number)).start()
-
             self.engine_info.status = MachineStatus.STARTED
             return ""
 
-    def _listen_command(self, listen_port, job_id):
-        address = ("localhost", int(listen_port))
-        listener = Listener(address, authkey="parent process secret password".encode())
-        conn = listener.accept()
-
-        while job_id in self.run_processes.keys():
-            clients = self.run_processes.get(job_id).get(RunProcessKey.PARTICIPANTS)
-            job_id = self.run_processes.get(job_id).get(RunProcessKey.JOB_ID)
-            try:
-                if conn.poll(0.1):
-                    received_data = conn.recv()
-                    command = received_data.get(ServerCommandKey.COMMAND)
-                    data = received_data.get(ServerCommandKey.DATA)
-
-                    if command == ServerCommandNames.GET_CLIENTS:
-                        return_data = {ServerCommandKey.CLIENTS: clients, ServerCommandKey.JOB_ID: job_id}
-                        conn.send(return_data)
-                    elif command == ServerCommandNames.AUX_SEND:
-                        targets = data.get("targets")
-                        topic = data.get("topic")
-                        request = data.get("request")
-                        timeout = data.get("timeout")
-                        fl_ctx = data.get("fl_ctx")
-                        replies = self.aux_send(
-                            targets=targets, topic=topic, request=request, timeout=timeout, fl_ctx=fl_ctx
-                        )
-                        conn.send(replies)
-                    elif command == ServerCommandNames.UPDATE_RUN_STATUS:
-                        execution_error = data.get("execution_error")
-                        if execution_error:
-                            with self.lock:
-                                run_process_info = self.run_processes.get(job_id)
-                                self.exception_run_processes[job_id] = run_process_info
-
-            except BaseException as e:
-                self.logger.warning(f"Failed to process the child process command: {secure_format_exception(e)}")
-
     def remove_exception_process(self, job_id):
         with self.lock:
             if job_id in self.exception_run_processes:
                 self.exception_run_processes.pop(job_id)
 
-    def wait_for_complete(self, job_id):
-        while True:
-            try:
-                self.send_command_to_child_runner_process(
-                    job_id=job_id, command_name=ServerCommandNames.HEARTBEAT, command_data={}, return_result=False
-                )
-                time.sleep(1.0)
-            except BaseException:
-                with self.lock:
-                    if job_id in self.run_processes:
-                        run_process_info = self.run_processes.pop(job_id)
-                        return_code = run_process_info[RunProcessKey.CHILD_PROCESS].poll()
-                        # if process exit but with Execution exception
-                        if return_code and return_code != 0:
-                            self.exception_run_processes[job_id] = run_process_info
-                self.engine_info.status = MachineStatus.STOPPED
-                break
+    def wait_for_complete(self, job_id, process):
+        process.wait()
+        with self.lock:
+            if job_id in self.run_processes:
+                run_process_info = self.run_processes.pop(job_id)
+                # return_code = run_process_info[RunProcessKey.CHILD_PROCESS].poll()
+                # # if process exit but with Execution exception
+                # if return_code and return_code != 0:
+                #     self.exception_run_processes[job_id] = run_process_info
+        self.engine_info.status = MachineStatus.STOPPED
 
     def _start_runner_process(
-        self, args, app_root, run_number, app_custom_folder, open_ports, job_id, job_clients, snapshot
+        self, args, app_root, run_number, app_custom_folder, open_ports, job_id, job_clients, snapshot, cell: Cell
     ):
         new_env = os.environ.copy()
         if app_custom_folder != "":
@@ -308,9 +240,9 @@ class ServerEngine(ServerEngineInternalSpec):
             + " -n "
             + str(run_number)
             + " -p "
-            + str(listen_port)
-            + " -c "
-            + str(open_ports[0])
+            + str(cell.get_internal_listener_url())
+            + " -u "
+            + str(cell.get_root_url_for_child())
             + " --set"
             + command_options
             + " print_conf=True restore_snapshot="
@@ -334,7 +266,7 @@ class ServerEngine(ServerEngineInternalSpec):
                 RunProcessKey.PARTICIPANTS: job_clients,
             }
 
-        threading.Thread(target=self.wait_for_complete, args=[run_number]).start()
+        threading.Thread(target=self.wait_for_complete, args=[run_number, process]).start()
         return process
 
     def get_job_clients(self, client_sites):
@@ -432,30 +364,8 @@ class ServerEngine(ServerEngineInternalSpec):
         else:
             return ""
 
-    def get_all_clients(self):
-        return list(self.server.client_manager.clients.keys())
-
     def get_client_from_name(self, client_name):
-        for c in self.get_clients():
-            if client_name == c.name:
-                return c
-        return None
-
-    def _get_all_clients_from_inputs(self, inputs):
-        clients = []
-        invalid_inputs = []
-        for item in inputs:
-            client = self.client_manager.clients.get(item)
-            # if item in self.get_all_clients():
-            if client:
-                clients.append(client)
-            else:
-                client = self.get_client_from_name(item)
-                if client:
-                    clients.append(client)
-                else:
-                    invalid_inputs.append(item)
-        return clients, invalid_inputs
+        return self.client_manager.get_client_from_name(client_name)
 
     def get_app_data(self, app_name: str) -> Tuple[str, object]:
         fullpath_src = os.path.join(self.server.admin_server.file_upload_dir, app_name)
@@ -555,7 +465,7 @@ class ServerEngine(ServerEngineInternalSpec):
     def send_aux_request(self, targets: [], topic: str, request: Shareable, timeout: float, fl_ctx: FLContext) -> dict:
         try:
             if not targets:
-                self.sync_clients_from_main_process()
+                # self.sync_clients_from_main_process()
                 targets = []
                 for t in self.get_clients():
                     targets.append(t.name)
@@ -569,102 +479,52 @@ class ServerEngine(ServerEngineInternalSpec):
             self.logger.error(f"Failed to send the aux_message: {topic} with exception: {secure_format_exception(e)}.")
 
     def sync_clients_from_main_process(self):
-        with self.parent_conn_lock:
-            data = {ServerCommandKey.COMMAND: ServerCommandNames.GET_CLIENTS, ServerCommandKey.DATA: {}}
-            self.parent_conn.send(data)
-            return_data = self.parent_conn.recv()
-            clients = return_data.get(ServerCommandKey.CLIENTS)
-            self.client_manager.clients = clients
-
-    def parent_aux_send(self, targets: [], topic: str, request: Shareable, timeout: float, fl_ctx: FLContext) -> dict:
-        with self.parent_conn_lock:
-            data = {
-                ServerCommandKey.COMMAND: ServerCommandNames.AUX_SEND,
-                ServerCommandKey.DATA: {
-                    "targets": targets,
-                    "topic": topic,
-                    "request": request,
-                    "timeout": timeout,
-                    "fl_ctx": get_serializable_data(fl_ctx),
-                },
-            }
-            self.parent_conn.send(data)
-            return_data = self.parent_conn.recv()
-            return return_data
+        return_data = self._get_clients_data(self.args.job_id)
+        clients = return_data.get(ServerCommandKey.CLIENTS)
+        self.client_manager.clients = clients
+
+    def _get_clients_data(self, job_id):
+        start = time.time()
+        request = new_cell_message({CellMessageHeaderKeys.JOB_ID: job_id}, fobs.dumps({}))
+        return_data = self.server.cell.send_request(
+            target=FQCN.ROOT_SERVER,
+            channel=CellChannel.SERVER_PARENT_LISTENER,
+            topic=ServerCommandNames.GET_CLIENTS,
+            request=request,
+        )
+        data = fobs.loads(return_data.payload)
+        if data.get(ServerCommandKey.CLIENTS):
+            return data
+        else:
+            if time.time() - start > self.kv_list.get("get_client_timeout", 10.0):
+                raise RuntimeError("Could not get the participants from parent process.")
+            time.sleep(1.0)
+            return self._get_clients_data(job_id)
 
     def update_job_run_status(self):
-        with self.parent_conn_lock:
-            with self.new_context() as fl_ctx:
-                execution_error = fl_ctx.get_prop(FLContextKey.FATAL_SYSTEM_ERROR, False)
-                data = {
-                    ServerCommandKey.COMMAND: ServerCommandNames.UPDATE_RUN_STATUS,
-                    ServerCommandKey.DATA: {
-                        "execution_error": execution_error,
-                    },
-                }
-                self.parent_conn.send(data)
-
-    def aux_send(self, targets: [], topic: str, request: Shareable, timeout: float, fl_ctx: FLContext) -> dict:
-        # Send the aux messages through admin_server
-        request.set_peer_props(fl_ctx.get_all_public_props())
-
-        message = Message(topic=ReservedTopic.AUX_COMMAND, body=fobs.dumps(request))
-        message.set_header(RequestHeader.JOB_ID, str(fl_ctx.get_prop(FLContextKey.CURRENT_RUN)))
-        requests = {}
-        for n in targets:
-            requests.update({n: message})
-
-        replies = self.server.admin_server.send_requests(requests, timeout_secs=timeout)
-        results = {}
-        for r in replies:
-            client_name = self.get_client_name_from_token(r.client_token)
-            if r.reply:
-                try:
-                    error_code = r.reply.get_header(MsgHeader.RETURN_CODE, ReturnCode.OK)
-                    if error_code != ReturnCode.OK:
-                        self.logger.error(f"Aux message send error: {error_code} from client: {client_name}")
-                        shareable = make_reply(ReturnCode.ERROR)
-                    else:
-                        shareable = fobs.loads(r.reply.body)
-                    results[client_name] = shareable
-                except BaseException as e:
-                    results[client_name] = make_reply(ReturnCode.COMMUNICATION_ERROR)
-                    self.logger.error(
-                        f"Received unexpected reply from client: {client_name}, "
-                        f"message body:{r.reply.body} processing topic:{topic} Error:{e}"
-                    )
-            else:
-                results[client_name] = None
-
-        return results
+        with self.new_context() as fl_ctx:
+            execution_error = fl_ctx.get_prop(FLContextKey.FATAL_SYSTEM_ERROR, False)
+            data = {"execution_error": execution_error}
+            job_id = fl_ctx.get_job_id()
+            request = new_cell_message({CellMessageHeaderKeys.JOB_ID: job_id}, fobs.dumps(data))
+            return_data = self.server.cell.fire_and_forget(
+                targets=FQCN.ROOT_SERVER,
+                channel=CellChannel.SERVER_PARENT_LISTENER,
+                topic=ServerCommandNames.UPDATE_RUN_STATUS,
+                message=request,
+            )
 
     def send_command_to_child_runner_process(self, job_id: str, command_name: str, command_data, return_result=True):
         result = None
         with self.lock:
-            command_conn = self.get_command_conn(job_id)
-            if command_conn:
-                data = {ServerCommandKey.COMMAND: command_name, ServerCommandKey.DATA: command_data}
-                command_conn.send(data)
-                if return_result:
-                    result = command_conn.recv()
-                    if result == NO_OP_REPLY:
-                        result = None
-        return result
+            fqcn = FQCN.join([FQCN.ROOT_SERVER, job_id])
+            request = new_cell_message({}, fobs.dumps(command_data))
+            return_data = self.server.cell.send_request(
+                target=fqcn, channel=CellChannel.SERVER_COMMAND, topic=command_name, request=request
+            )
+            result = fobs.loads(return_data.payload)
 
-    def get_command_conn(self, job_id):
-        # this function need to be called with self.lock
-        port = self.run_processes.get(job_id, {}).get(RunProcessKey.LISTEN_PORT)
-        command_conn = self.run_processes.get(job_id, {}).get(RunProcessKey.CONNECTION)
-
-        if not command_conn:
-            try:
-                address = ("localhost", port)
-                command_conn = CommandClient(address, authkey="client process secret password".encode())
-                command_conn = ClientConnection(command_conn)
-                self.run_processes[job_id][RunProcessKey.CONNECTION] = command_conn
-            except Exception:
-                pass
-        return command_conn
+        return result
 
     def persist_components(self, fl_ctx: FLContext, completed: bool):
         self.logger.info("Start saving snapshot on server.")
@@ -693,13 +553,10 @@ class ServerEngine(ServerEngineInternalSpec):
 
             job_info = fl_ctx.get_prop(FLContextKey.JOB_INFO)
             if not job_info:
-                with self.parent_conn_lock:
-                    data = {ServerCommandKey.COMMAND: ServerCommandNames.GET_CLIENTS, ServerCommandKey.DATA: {}}
-                    self.parent_conn.send(data)
-                    return_data = self.parent_conn.recv()
-                    job_id = return_data.get(ServerCommandKey.JOB_ID)
-                    job_clients = return_data.get(ServerCommandKey.CLIENTS)
-                    fl_ctx.set_prop(FLContextKey.JOB_INFO, (job_id, job_clients))
+                return_data = self._get_clients_data(job_id)
+                job_id = return_data.get(ServerCommandKey.JOB_ID)
+                job_clients = return_data.get(ServerCommandKey.CLIENTS)
+                fl_ctx.set_prop(FLContextKey.JOB_INFO, (job_id, job_clients))
             else:
                 (job_id, job_clients) = job_info
             snapshot.set_component_snapshot(
@@ -757,6 +614,19 @@ class ServerEngine(ServerEngineInternalSpec):
             errors = {}
         return errors
 
+    def reset_errors(self, job_id) -> str:
+        errors = None
+        try:
+            self.send_command_to_child_runner_process(
+                job_id=job_id,
+                command_name=ServerCommandNames.RESET_ERRORS,
+                command_data={},
+            )
+        except BaseException:
+            self.logger.error(f"Failed to reset_errors for JOB: {job_id}")
+
+        return f"reset the server error stats for job: {job_id}"
+
     def _send_admin_requests(self, requests, timeout_secs=10) -> List[ClientReply]:
         return self.server.admin_server.send_requests(requests, timeout_secs=timeout_secs)
 
@@ -837,10 +707,10 @@ def server_shutdown(server, touch_file):
         os.utime(touch_file, None)
 
     try:
-        server.fl_shutdown(shutdown_period=30.0)
-        server.admin_server.stop()
-        time.sleep(3.0)
+        # server.admin_server.stop()
+        server.fl_shutdown()
+        # time.sleep(3.0)
     finally:
         security_close()
         server.status = ServerStatus.SHUTDOWN
-        sys.exit(2)
+        # sys.exit(2)
